### ISSUE TITLE
tril and triu return value has the wrong dtype (migrated from Trac #1848)

### ISSUE THREAD

**Issue by thouis** [thouis/numpy-trac-migration#3396]
Original ticket http://projects.scipy.org/numpy/ticket/1848
Reported 2011-05-28 by atmention:WarrenWeckesser, assigned to unknown.

tril (and triu, which defers to tril for documentation of its
return value), says that the return value is the same type as
the input.  This fails for float32 and complex64 input arrays:
    In [16]: a = array([[1,2],[3,4]], dtype=float32)

```
In [17]: tril(a).dtype
Out[17]: dtype('float64')

In [18]: b = array([[1+1j,2],[3,4]], dtype=complex64)

In [19]: tril(b).dtype
Out[19]: dtype('complex128')
```

tril looks like this:
    def tril(m, k=0):
        m = asanyarray(m)
        out = multiply(tri(m.shape[0], m.shape[1], k=k, dtype=int),m)
        return out
and `tri(m.shape[0], m.shape[1], k=k, dtype=int)` is an array of
integer zeros and ones.  The problem is that the multiplication
of a float32 array by an array of integers result in a float64
array:
    In [168]: a = array([[1.0, 2.0], [3.0, 4.0]], dtype=float32)

```
In [169]: m = array([[1, 1], [0, 1]])

In [170]: a*m
Out[170]: 
array([[ 1.,  2.],
       [ 0.,  4.]])

In [171]: m*a
Out[171]: 
array([[ 1.,  2.],
       [ 0.,  4.]])
```

Note that multiplying a float32 array by an integer scalar does not 
change the type:
    In [172]: a*3
    Out[172]: 
    array([[  3.,   6.],
           [  9.,  12.]], dtype=float32)

```
In [173]: a*array(3)
Out[173]: 
array([[  3.,   6.],
       [  9.,  12.]], dtype=float32)
```

which seems inconsistent.

The problem with tril and triu was also pointed out by "eat" in #1723.

**Comment by thouis** [comment]
Comment in Trac by atmention:WarrenWeckesser, 2011-05-28

The problem was reported earlier in a scipy ticket: http://projects.scipy.org/scipy/ticket/1449

**Comment by thouis** [comment]
Comment in Trac by atmention:WarrenWeckesser, 2011-05-28

**Comment by thouis** [comment]
Comment in Trac by trac user Peter17, 2011-06-20

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2011-08-31

Fixed in commit:1af2f37

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2012-04-30
